casper/src/main/scala/coop/rchain/casper/engine/Initializing.scala

```diff
@@ -38,121 +47,178 @@ class Initializing[F[_]
               _ <- Log[F].info(
-                    s"Valid Last Finalized Block received ${PrettyPrinter.buildString(lastFinalizedBlock)}"
+                    s"Valid Last Finalized Block received ${PrettyPrinter.buildString(approvedBlock)}"
                   )
               _ <- Log[F].warn(
-                    s"JUSTIF: ${lastFinalizedBlock.justifications.map(_.latestBlockHash).map(PrettyPrinter.buildString).mkString(", ")}"
+                    s"JUSTIF: ${approvedBlock.justifications.map(_.latestBlockHash).map(PrettyPrinter.buildString).mkString(", ")}"
                   )
 
               _ <- EventLog[F].publish(
                     shared.Event.ApprovedBlockReceived(
                       PrettyPrinter
-                        .buildStringNoLimit(lastFinalizedBlock.blockHash)
+                        .buildStringNoLimit(approvedBlock.blockHash)
                     )
                   )
 
-              // TODO: resolve validation of received last finalized block
               // Add last finalized block as ApprovedBlock
-              approvedBlock = ApprovedBlock(ApprovedBlockCandidate(lastFinalizedBlock, 0), Nil)
-              _             <- insertIntoBlockAndDagStore[F](lastFinalizedBlock, approvedBlock)
-              _             <- LastApprovedBlock[F].set(approvedBlock)
+              ab = ApprovedBlock(ApprovedBlockCandidate(approvedBlock, 0), Nil)
+              _  <- insertIntoBlockAndDagStore[F](approvedBlock, ab)
+              _  <- LastApprovedBlock[F].set(ab)
 
               // Update last finalized block with received block hash
-              _ <- LastFinalizedStorage[F].put(lastFinalizedBlock.blockHash)
+              _ <- LastFinalizedStorage[F].put(approvedBlock.blockHash)
 
               // Transition to restore last finalized state
-              _ <- transitionToLastFinalizedState(
-                    shardId,
-                    finalizationRate,
-                    validatorId
-                  )
+              _ <- requestApprovedState
             } yield ()
           } else
             Log[F].info("Invalid LastFinalizedBlock received; refusing to add.")
     } yield ()
   }
 
+  def requestApprovedState: F[Unit] = {
+
+    import cats.instances.list._
+    import coop.rchain.catscontrib.Catscontrib.ToBooleanF
+    def addBlockToDag(block: BlockMessage, approvedBlock: ApprovedBlock): F[Unit] =
+      for {
+        dag   <- BlockDagStorage[F].getRepresentation
+        added <- dag.contains(block.blockHash)
+        _ <- BlockDagStorage[F]
+              .insert(
+                block,
+                approvedBlock.candidate.block,
+                invalid = false
+              )
+              .unlessA(added)
+        missingDeps <- ProtoUtil.dependenciesHashesOf(block).filterA(dag.contains(_).not)
+        blocksToAdd <- missingDeps.traverse(BlockStore[F].get)
+        sortedDeps = blocksToAdd.flatten.sortWith(
+          _.body.state.blockNumber > _.body.state.blockNumber
+        )
```

Comment:
Because `Ordering[Long]` is defined there is more convenient function `sortBy` which only need selection of the field.
`blocksToAdd.flatten.sortBy(_.body.state.blockNumber)`